File: app/src/main/java/com/gytxtx/openjbd/BmsDashboardStore.java
package com.gytxtx.openjbd;

import java.util.ArrayList;
import java.util.List;

final class BmsDashboardStore {
    private static final List<Listener> LISTENERS = new ArrayList<>();
    private static Snapshot snapshot;

    private BmsDashboardStore() {
    }

    static synchronized Snapshot getSnapshot() {
        return snapshot;
    }

    static void update(Snapshot value) {
        List<Listener> listeners;
        synchronized (BmsDashboardStore.class) {
            snapshot = value;
            listeners = new ArrayList<>(LISTENERS);
        }
        for (Listener listener : listeners) {
            listener.onDashboardSnapshotChanged(value);
        }
    }

    static synchronized void addListener(Listener listener) {
        if (!LISTENERS.contains(listener)) {
            LISTENERS.add(listener);
        }
    }

    static synchronized void removeListener(Listener listener) {
        LISTENERS.remove(listener);
    }

    interface Listener {
        void onDashboardSnapshotChanged(Snapshot snapshot);
    }

    static final class Snapshot {
        final int soc;
        final float voltage;
        final float current;
        final float power;

        Snapshot(int soc, float voltage, float current, float power) {
            this.soc = soc;
            this.voltage = voltage;
            this.current = current;
            this.power = power;
        }
    }
}

File: app/src/main/kotlin/com/gytxtx/openjbd/BmsDashboardStore.kt
package com.gytxtx.openjbd

object BmsDashboardStore {
    @Volatile
    private var _snapshot: Snapshot? = null
    private val listeners = mutableListOf<Listener>()

    @JvmStatic
    @Synchronized
    fun getSnapshot(): Snapshot? = _snapshot

    @JvmStatic
    fun update(value: Snapshot?) {
        val copy: List<Listener>
        synchronized(this) {
            _snapshot = value
            copy = listeners.toList()
        }
        for (listener in copy) {
            listener.onDashboardSnapshotChanged(value)
        }
    }

    @JvmStatic
    @Synchronized
    fun addListener(listener: Listener) {
        if (!listeners.contains(listener)) {
            listeners.add(listener)
        }
    }

    @JvmStatic
    @Synchronized
    fun removeListener(listener: Listener) {
        listeners.remove(listener)
    }

    interface Listener {
        fun onDashboardSnapshotChanged(snapshot: Snapshot?)
    }

    class Snapshot(
        @JvmField val soc: Int,
        @JvmField val voltage: Float,
        @JvmField val current: Float,
        @JvmField val power: Float
    )
}
